drivers/network/dd/e1000/hardware.c

```diff
@@ -0,0 +1,817 @@
+    {
+        if (!E1000ReadEeprom(Adapter, (UCHAR)n, &AddrWord))
+            return NDIS_STATUS_FAILURE;
+        Adapter->PermanentMacAddress[n * 2 + 0] = AddrWord & 0xff;
+        Adapter->PermanentMacAddress[n * 2 + 1] = (AddrWord >> 8) & 0xff;
+    }
+
+    NDIS_DbgPrint(MIN_TRACE, ("MAC: %02x:%02x:%02x:%02x:%02x:%02x\n",
+                              Adapter->PermanentMacAddress[0],
+                              Adapter->PermanentMacAddress[1],
+                              Adapter->PermanentMacAddress[2],
+                              Adapter->PermanentMacAddress[3],
+                              Adapter->PermanentMacAddress[4],
+                              Adapter->PermanentMacAddress[5]));
+    return NDIS_STATUS_SUCCESS;
+}
+
+NDIS_STATUS
+NTAPI
+NICUpdateMulticastList(
+    IN PE1000_ADAPTER Adapter)
+{
+    UINT n;
+    NDIS_DbgPrint(MAX_TRACE, ("Called.\n"));
+
+    for (n = 0; n < MAXIMUM_MULTICAST_ADDRESSES; ++n)
+    {
+        ULONG Ral = *(ULONG *)Adapter->MulticastList[n].MacAddress;
+        ULONG Rah = *(USHORT *)&Adapter->MulticastList[n].MacAddress[4];
+
+        if (Rah || Ral)
+        {
+            Rah |= E1000_RAH_AV;
+
+            E1000WriteUlong(Adapter, E1000_REG_RAL + (8*n), Ral);
+            E1000WriteUlong(Adapter, E1000_REG_RAH + (8*n), Rah);
+        }
+        else
+        {
+            E1000WriteUlong(Adapter, E1000_REG_RAH + (8*n), 0);
+            E1000WriteUlong(Adapter, E1000_REG_RAL + (8*n), 0);
+        }
+    }
+
+    return NDIS_STATUS_SUCCESS;
+}
+
+NDIS_STATUS
+NTAPI
+NICApplyPacketFilter(
+    IN PE1000_ADAPTER Adapter)
+{
+    ULONG FilterMask;
+
+    E1000ReadUlong(Adapter, E1000_REG_RCTL, &FilterMask);
+
+    FilterMask &= ~E1000_RCTL_FILTER_BITS;
+    FilterMask |= PacketFilterToMask(Adapter->PacketFilter);
+    E1000WriteUlong(Adapter, E1000_REG_RCTL, FilterMask);
+
+    return NDIS_STATUS_SUCCESS;
+}
+
+NDIS_STATUS
+NTAPI
+NICApplyInterruptMask(
+    IN PE1000_ADAPTER Adapter)
+{
+    NDIS_DbgPrint(MAX_TRACE, ("Called.\n"));
+
+    E1000WriteUlong(Adapter, E1000_REG_IMS, Adapter->InterruptMask /*| 0x1F6DC*/);
+    return NDIS_STATUS_SUCCESS;
+}
+
+NDIS_STATUS
+NTAPI
+NICDisableInterrupts(
+    IN PE1000_ADAPTER Adapter)
+{
+    NDIS_DbgPrint(MAX_TRACE, ("Called.\n"));
+
+    E1000WriteUlong(Adapter, E1000_REG_IMC, ~0);
+    return NDIS_STATUS_SUCCESS;
+}
+
+ULONG
+NTAPI
+NICInterruptRecognized(
+    IN PE1000_ADAPTER Adapter,
+    OUT PBOOLEAN InterruptRecognized)
+{
+    ULONG Value;
+
+    //NDIS_DbgPrint(MAX_TRACE, ("Called.\n"));
+
+    /* Reading the interrupt acknowledges them */
+    E1000ReadUlong(Adapter, E1000_REG_ICR, &Value);
+
+    *InterruptRecognized = (Value & Adapter->InterruptMask) != 0;
+
+    NDIS_DbgPrint(MAX_TRACE, ("NICInterruptRecognized(0x%x, 0x%x).\n", Value, *InterruptRecognized));
+
+    return (Value & Adapter->InterruptMask);
+}
+
+VOID
+NTAPI
+NICUpdateLinkStatus(
+    IN PE1000_ADAPTER Adapter)
+{
+    UINT32 DeviceStatus;
```

Comment:
../drivers/network/dd/e1000/hardware.c:780:5: error: passing argument 3 of 'E1000ReadUlong' from incompatible pointer type [-Werror]
../drivers/network/dd/e1000/hardware.c:34:12: note: expected 'PULONG' but argument is of type 'UINT32 *'